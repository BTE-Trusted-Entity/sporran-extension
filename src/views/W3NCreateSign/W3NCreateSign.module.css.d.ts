export const __esModule: true;
export const container: string;
export const heading: string;
export const subline: string;
export const didLine: string;
export const costs: string;
export const details: string;
export const label: string;
export const errorTooltip: string;
export const buttonsLine: string;
export const next: string;
export const back: string;
