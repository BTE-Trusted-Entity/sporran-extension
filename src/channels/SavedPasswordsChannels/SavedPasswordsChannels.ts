import { BrowserChannel } from '../base/BrowserChannel/BrowserChannel';
import { channelsEnum } from '../base/channelsEnum';

interface SavedPassword {
  password: string;
  timestamp: number;
}

const savedPasswords: Record<string, SavedPassword> = {};

interface SavePasswordInput {
  password: string;
  address: string;
}

export const savePasswordChannel = new BrowserChannel<SavePasswordInput>(
  channelsEnum.savePassword,
);

export async function savePassword({
  address,
  password,
}: SavePasswordInput): Promise<void> {
  savedPasswords[address] = { password, timestamp: Date.now() };
}

export const getPasswordChannel = new BrowserChannel<
  string,
  string | undefined
>(channelsEnum.getPassword);

export async function getPassword(
  address: string,
): Promise<string | undefined> {
  return savedPasswords[address]?.password;
}

export const forgetPasswordChannel = new BrowserChannel<string>(
  channelsEnum.forgetPassword,
);

export async function forgetPassword(address: string): Promise<void> {
  delete savedPasswords[address];
}

export const hasSavedPasswordsChannel = new BrowserChannel<void, boolean>(
  channelsEnum.hasSavedPasswords,
);

export async function hasSavedPasswords(): Promise<boolean> {
  return Object.values(savedPasswords).length > 0;
}

export const forgetAllPasswordsChannel = new BrowserChannel(
  channelsEnum.forgetAllPasswords,
);

export async function forgetAllPasswords(): Promise<void> {
  for (const password in savedPasswords) {
    delete savedPasswords[password];
  }
}

const saveDuration = 15 * 60 * 1000;
const intervalDuration = 1 * 60 * 1000;

function checkExpiredPasswords(): void {
  const oldestPossible = Date.now() - saveDuration;
  for (const password in savedPasswords) {
    const { timestamp } = savedPasswords[password];
    if (timestamp < oldestPossible) {
      delete savedPasswords[password];
    }
  }
}

export function schedulePasswordsCheck(): void {
  setInterval(checkExpiredPasswords, intervalDuration);
}
