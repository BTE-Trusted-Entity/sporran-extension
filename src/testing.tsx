export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
