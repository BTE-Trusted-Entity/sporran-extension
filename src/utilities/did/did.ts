import {
  ConfigService,
  Did,
  DidDocument,
  DidEncryptionKey,
  DidUri,
  Utils,
} from '@kiltprotocol/sdk-js';

import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

export function isFullDid(did: DidUri | undefined): boolean {
  if (!did) {
    // could be a legacy identity without DID
    return false;
  }
  return Did.parse(did).type === 'full';
}

export async function getFullDidDocument(did: DidUri): Promise<DidDocument> {
  const api = ConfigService.get('api');
  return Did.linkedInfoFromChain(await api.call.did.query(Did.toChain(did)))
    .document;
}

export function useFullDidDocument(did: DidUri): DidDocument | undefined {
  return useAsyncValue(getFullDidDocument, [did]);
}

export async function getDidDocument(did: DidUri): Promise<DidDocument> {
  return isFullDid(did)
    ? await getFullDidDocument(did)
    : Did.parseDocumentFromLightDid(did);
}

export function getDidEncryptionKey({
  keyAgreement,
}: DidDocument): DidEncryptionKey {
  if (!keyAgreement || !keyAgreement[0]) {
    throw new Error('encryptionKey is not defined somehow');
  }
  return keyAgreement[0];
}

export function parseDidUri(did: DidUri): ReturnType<typeof Did.parse> & {
  fullDid: DidUri;
} {
  const parsed = Did.parse(did);
  const fullDid = parsed.type === 'full' ? did : Did.getFullDidUri(did);

  return {
    ...parsed,
    fullDid,
  };
}

export async function needLegacyDidCrypto(
  did: DidUri | undefined,
): Promise<boolean> {
  if (!did) {
    // DID was deactivated, no action needed.
    return false;
  }
  if (!isFullDid(did)) {
    return false;
  }

  try {
    const encryptionKey = getDidEncryptionKey(await getDidDocument(did));
    return (
      Utils.Crypto.u8aToHex(encryptionKey.publicKey) ===
      '0xf2c90875e0630bd1700412341e5e9339a57d2fefdbba08de1cac8db5b4145f6e'
    );
  } catch {
    // getDidDocument might throw if the DID is not on-chain anymore (removed, another endpoint),
    // no legacy crypto needed in that case
    return false;
  }
}
