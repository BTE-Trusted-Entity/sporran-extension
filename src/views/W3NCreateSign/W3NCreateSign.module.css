.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.subline {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0 0 0.8125rem;
}

.didLine {
  composes: didLine from '../../components/CopyValue/CopyValue.module.css';
  margin-top: 1.5rem;
}

.costs {
  composes: smallText from '../../components/Typography/Typography.module.css';
  font-size: 1.25rem;
  margin: 0.5rem 0 0;
  display: flex;
  align-items: center;
  justify-content: center;
}

.details {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0.5rem 0 0;
}

.label {
  color: var(--color-grey-medium);
}

.errorTooltip {
  composes: tooltip from '../../components/ErrorTooltip/ErrorTooltip.module.css';
}

.buttonsLine {
  composes: buttonsLine from '../../components/Button/Button.module.css';
}

.next {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.back {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}
