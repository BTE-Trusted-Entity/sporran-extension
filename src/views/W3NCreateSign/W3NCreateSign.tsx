import { FormEvent, useCallback, useRef } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { browser } from 'webextension-polyfill-ts';

import { ConfigService, Did } from '@kiltprotocol/sdk-js';

import * as styles from './W3NCreateSign.module.css';

import { Identity } from '../../utilities/identities/types';
import { IdentitySlide } from '../../components/IdentitySlide/IdentitySlide';
import { CopyValue } from '../../components/CopyValue/CopyValue';
import { LinkBack } from '../../components/LinkBack/LinkBack';
import { Stats } from '../../components/Stats/Stats';
import {
  PasswordField,
  usePasswordField,
} from '../../components/PasswordField/PasswordField';
import {
  getIdentityCryptoFromSeed,
  getIdentityDid,
} from '../../utilities/identities/identities';
import { TxStatusModal } from '../../components/TxStatusModal/TxStatusModal';
import {
  asKiltCoins,
  KiltAmount,
} from '../../components/KiltAmount/KiltAmount';
import { KiltCurrency } from '../../components/KiltCurrency/KiltCurrency';
import { ExplainerModal } from '../../components/ExplainerModal/ExplainerModal';
import { useSubmitStates } from '../../utilities/useSubmitStates/useSubmitStates';
import { useKiltCosts } from '../../utilities/w3nCreate/w3nCreate';

interface Props {
  identity: Identity;
}

export function W3NCreateSign({ identity }: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const history = useHistory();
  const { goBack } = history;

  const { web3name } = useParams() as { web3name: string };

  const { address } = identity;
  const did = getIdentityDid(identity);

  const { deposit, fee, total, insufficientKilt } = useKiltCosts(address, did);

  const { submit, modalProps, submitting } = useSubmitStates();

  const passwordField = usePasswordField();
  const handleSubmit = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();

      const { keypair, seed } = await passwordField.get(event);
      const { sign } = await getIdentityCryptoFromSeed(seed);

      const api = ConfigService.get('api');
      const authorized = await Did.authorizeTx(
        did,
        api.tx.web3Names.claim(web3name),
        sign,
        keypair.address,
      );
      await submit(keypair, authorized);
    },
    [did, passwordField, submit, web3name],
  );

  const portalRef = useRef<HTMLDivElement>(null);

  if (!deposit || !fee || !total) {
    return null; // blockchain data pending
  }

  return (
    <form className={styles.container} onSubmit={handleSubmit}>
      <h1 className={styles.heading}>{t('view_W3NCreateSign_heading')}</h1>
      <p className={styles.subline}>{t('view_W3NCreateSign_subline')}</p>

      <IdentitySlide identity={identity} />

      <CopyValue value={did} label="DID" className={styles.didLine} />

      <p className={styles.costs}>
        {t('view_W3NCreateSign_total')}
        <KiltAmount amount={total} type="costs" smallDecimals />
      </p>
      <p className={styles.details}>
        <ExplainerModal portalRef={portalRef}>
          {t('view_W3NCreateSign_explainer')}
        </ExplainerModal>
        {t('view_W3NCreateSign_deposit')}
        {asKiltCoins(deposit, 'costs')} <KiltCurrency />
        {t('view_W3NCreateSign_fee')}
        {asKiltCoins(fee, 'costs')} <KiltCurrency />
      </p>
      <p className={styles.details}>
        <span className={styles.label}>{t('view_W3NCreateSign_label')}</span>
        {web3name}
      </p>

      <PasswordField identity={identity} autoFocus password={passwordField} />

      <p className={styles.buttonsLine}>
        <button type="button" onClick={goBack} className={styles.back}>
          {t('common_action_back')}
        </button>

        <button
          type="submit"
          className={styles.next}
          disabled={insufficientKilt || submitting}
        >
          {t('common_action_sign')}
        </button>

        <output
          className={styles.errorTooltip}
          hidden={!insufficientKilt || Boolean(modalProps)}
        >
          {t('view_W3NCreateSign_insufficientFunds', [
            asKiltCoins(total, 'costs'),
          ])}
        </output>
      </p>

      {modalProps && (
        <TxStatusModal
          {...modalProps}
          identity={identity}
          messages={{
            pending: t('view_W3NCreateSign_pending'),
            success: t('view_W3NCreateSign_success'),
            error: t('view_W3NCreateSign_error'),
          }}
        />
      )}

      <div ref={portalRef} />

      <LinkBack />
      <Stats />
    </form>
  );
}
