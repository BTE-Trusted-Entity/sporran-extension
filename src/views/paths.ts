export { generatePath } from 'react-router-dom';

export const paths = {
  home: '/',
  settings: '/settings',
  access: '/access',
  identity: {
    base: '/identity/',
    add: '/identity/add',
    create: {
      start: '/identity/create',
      backup: '/identity/create/backup',
      verify: '/identity/create/verify',
      password: '/identity/create/password',
    },
    import: {
      start: '/identity/import',
      password: '/identity/import/password',
    },

    overview: '/identity/:address/:type(created|imported|pwreset)?',
    reset: {
      start: '/identity/:address/reset',
      password: '/identity/:address/reset/password',
    },
    send: {
      start: '/identity/:address/send',
      review: '/identity/:address/send/review',
      warning: '/identity/:address/send/warning',
    },
    receive: '/identity/:address/receive',
    credentials: {
      base: '/identity/:address/credentials',
      presentation: '/identity/:address/credentials/:hash/presentation',
    },
    remove: '/identity/:address/remove',
    did: {
      upgrade: {
        start: '/identity/:address/did/upgrade',
        kilt: '/identity/:address/did/upgrade/kilt',
        euro: '/identity/:address/did/upgrade/euro',
      },
      manage: {
        start: '/identity/:address/did',
        downgrade: {
          base: '/identity/:address/did/downgrade',
          warning: {
            web3name: '/identity/:address/did/downgrade/warning/web3name',
            credentials: '/identity/:address/did/downgrade/warning/credentials',
          },
          sign: '/identity/:address/did/downgrade/sign',
        },
        endpoints: {
          start: '/identity/:address/did/endpoints',
          edit: '/identity/:address/did/endpoints/:id?',
          sign: '/identity/:address/did/endpoints/sign',
        },
      },
      repair: '/identity/:address/did/repair',
    },
    web3name: {
      create: {
        base: '/identity/:address/web3name/create',
        info: '/identity/:address/web3name/create/info',
        form: '/identity/:address/web3name/create/form',
        choose: '/identity/:address/web3name/create/choose/:web3name',
        kilt: '/identity/:address/web3name/create/kilt/:web3name',
        euro: '/identity/:address/web3name/create/euro/:web3name',
      },
      manage: {
        start: '/identity/:address/web3name',
        remove: '/identity/:address/web3name/remove',
      },
    },
  },
  popup: {
    base: '/popup',
    import: '/popup/import',
    access: '/popup/access',
    claim: '/identity/:address/claim',
    save: '/popup/save',
    share: {
      start: '/popup/share',
      sign: '/popup/share/sign',
    },
    sign: '/popup/sign',
    signRaw: '/popup/signRaw',
    signDid: {
      start: '/identity/:address/did/sign',
      sign: '/identity/:address/did/sign/password',
      credentials: '/identity/:address/did/sign/credentials',
    },
    signDidExtrinsic: '/identity/:address/did/signExtrinsic',
    createDid: '/identity/:address/did/create',
  },
};
