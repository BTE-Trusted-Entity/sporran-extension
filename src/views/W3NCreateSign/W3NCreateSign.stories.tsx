import { Meta } from '@storybook/react';

import { MemoryRouter, Route } from 'react-router-dom';

import { identitiesMock } from '../../utilities/identities/IdentitiesProvider.mock';

import { generatePath, paths } from '../paths';

import { W3NCreateSign } from './W3NCreateSign';

export default {
  title: 'Views/W3NCreateSign',
  component: W3NCreateSign,
} as Meta;

const identity =
  identitiesMock['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo'];

export function Template(): JSX.Element {
  return (
    <MemoryRouter
      initialEntries={[
        generatePath(paths.identity.web3name.create.kilt, {
          address: 'FOO',
          web3name: 'fancy-name',
        }),
      ]}
    >
      <Route path={paths.identity.web3name.create.kilt}>
        <W3NCreateSign identity={identity} />
      </Route>
    </MemoryRouter>
  );
}
