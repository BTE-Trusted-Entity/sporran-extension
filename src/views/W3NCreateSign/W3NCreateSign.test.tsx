import { BalanceUtils, DidDocument } from '@kiltprotocol/sdk-js';

import { MemoryRouter, Route } from 'react-router-dom';

import { identitiesMock, render } from '../../testing/testing';
import { waitForGetPassword } from '../../channels/SavedPasswordsChannels/SavedPasswordsChannels.mock';
import '../../components/useCopyButton/useCopyButton.mock';
import { useAsyncValue } from '../../utilities/useAsyncValue/useAsyncValue';
import { useDepositWeb3Name } from '../../utilities/getDeposit/getDeposit';
import { useFullDidDocument } from '../../utilities/did/did';

import { generatePath, paths } from '../paths';

import { W3NCreateSign } from './W3NCreateSign';

jest.mock('../../utilities/useAsyncValue/useAsyncValue');
jest.mocked(useAsyncValue).mockReturnValue(BalanceUtils.toFemtoKilt(0.01));

jest.mock('../../utilities/getDeposit/getDeposit');
jest
  .mocked(useDepositWeb3Name)
  .mockReturnValue({ amount: BalanceUtils.toFemtoKilt(2) });

jest.mock('../../utilities/did/did');
jest.mocked(useFullDidDocument).mockReturnValue({} as DidDocument);

const identity =
  identitiesMock['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo'];

describe('W3NCreateSign', () => {
  it('should match the snapshot', async () => {
    const { container } = render(
      <MemoryRouter
        initialEntries={[
          generatePath(paths.identity.web3name.create.kilt, {
            address: 'FOO',
            web3name: 'fancy-name',
          }),
        ]}
      >
        <Route path={paths.identity.web3name.create.kilt}>
          <W3NCreateSign identity={identity} />
        </Route>
      </MemoryRouter>,
    );
    await waitForGetPassword();
    expect(container).toMatchSnapshot();
  });
});
