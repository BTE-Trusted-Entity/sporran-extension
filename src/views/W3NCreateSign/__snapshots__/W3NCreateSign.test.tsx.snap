// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`W3NCreateSign should match the snapshot 1`] = `
<div>
  <form
    class="container"
  >
    <h1
      class="heading"
    >
      web3name
    </h1>
    <p
      class="subline"
    >
      Connect a web3name to your DID
    </p>
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="didLine"
    >
      <input
        aria-label="DID"
        class="long"
        readonly=""
        value="did:kilt:4oeJ76hdj84xnwCNqijUHUCTmfwXgSZ4vmxLEiTEYgQdBCcZ"
      />
      <button
        aria-label="copy to clipboard"
        class="copy"
        title="copy to clipboard"
        type="button"
      />
    </p>
    <p
      class="costs"
    >
      Total cost: 
      <span
        class="amount"
      >
        2
        <span
          class="fractional"
        >
          .0100
        </span>
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </p>
    <p
      class="details"
    >
      <button
        aria-label="Information"
        class="show"
        type="button"
      />
       Deposit 
      2.0000
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
       + fee 
      0.0100
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
    </p>
    <p
      class="details"
    >
      <span
        class="label"
      >
        web3name: 
      </span>
      fancy-name
    </p>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <button
        class="back"
        type="button"
      >
        Back
      </button>
      <button
        class="next"
        disabled=""
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
      >
        You need at least 2.0100 KILTs to create the web3name
      </output>
    </p>
    <div />
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </form>
</div>
`;
